rts/motoko-rts/src/stabilization/metadata.rs

```diff
@@ -0,0 +1,224 @@
+//! Graph-copy-bazed serialization format:
+//!
+//! (Very first word is zeroed and backed up in last page)
+//! -- Stable memory
+//! Raw stable memory or region data, size S in pages
+//! -- Stable variables
+//! Serialized data address N:
+//!   Serialized object graph, length L
+//!   (possible zero padding)
+//! Type descriptor address M:
+//!   Candid type table
+//!     Byte length (u32)
+//!     Data
+//!   Type offset table
+//!     Byte length (u32)
+//!     Data
+//!   (possible zero padding)
+//! -- Last physical page (metadata):
+//!   (zero padding to align at page end)
```

Comment:
Perhaps we could store the instruction count of serialization here, and then use it to compute the cost of upgrade to store and retrieve from a global, if we want such stats.

Reply:
Great idea, this would simplify upgrade benchmarking.

Reply:
**Update**: I now record the serialization costs and offer `Prim.rts_upgrade_instructions()` to retrieve the total number of instructions of the last upgrade (serialization plus deserialization).